Move the remove icon button to the ellipsis menu as a menu button

talldan opened the pull request:
## Description
Fixes #7216, moves the remove icon button to the ellipsis menu as a menu button.

## How has this been tested?
- Added unit tests for 'Remove Block' menu button
- Performed manual tests including removing single and multiple blocks

## Screenshots
![screen shot 2018-06-26 at 11 24 08 am](https://user-images.githubusercontent.com/677833/41887506-8854fd5c-7933-11e8-9928-5a7d31425715.png)


## Types of changes
- Removed old icon button version of remove button
- Altered existing Remove menu button (which still seemed to be present from previous version) to display like other menu buttons
- Added tests for menu button

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards
- [x] My code has proper inline documentation.

talldan committed:
Remove BlockRemoveButton from BlockSettingsMenu

talldan committed:
Add existing BlockRemoveButton to BlockSettingsMenu

talldan committed:
Update BlockSettingsMenu BlockRemoveButton appearance

Add the label 'Remove Block' to the BlockRemoveButton and change the classname
to editor-block-settings-menu__control so the the icon button is consistent
with other menu controls

talldan committed:
Add unit tests for block-remove-button

karmatosed commented:
Thanks for working on this @noisysocks. Could you add a seperator using the style used in other drop downs?

![artboard](https://user-images.githubusercontent.com/253067/41926612-c65812e0-7967-11e8-8388-183126b7b894.png)

This visually creates a good divide in actions.

talldan committed:
Add MenuGroup components to separate menu items by their action

Intiially this just separates other menu items from the
'Remove block' menu item. This commit is a stepping stone to
separating both Remove Block and Delete Shared Block from the
other menu items.

talldan committed:
Apply border between MenuGroups

Previously, this border between menu groups was only being applied to
the MenuGroups in the Edit Post More Menu. This moves it to the MenuGroup component
so that it is applied to all MenuGroups

talldan committed:
Extract button for deleting shared block into its own component

Extract the button into a new SharedBlockDeleteButton, so that it can be
rendered separately in a menu group with the RemoveBlock button.

talldan committed:
Extract test for SharedBlockDeleteButton from SharedBlockSettings

Refactor tests to match refactoring to components. Add a snapshot test
for SharedBlockDeleteButton

talldan commented:
Well, I thought adding a separator would be nice and easy, but there were a few intricacies:

- In the other menu, a `MenuGroup` component is used to separate the menu items. The styling was applied only to that menu's `MenuGroup`s though. I've changed it now so that it applied to all `MenuGroup`s. I had a search in the codebase, and the only thing that it might affect that I'm unsure how to view is something called `PluginSidebarMoreMenuItem` - I guess I'll need some plugins to view that?
- I had to refactor a bit, since the menu items for 'Convert to Regular Block' and 'Delete Shared Block' were actually rendered in one parent component and because of that it was impossible to get a `MenuGroup` around the menu items we wanted. I've extracted the Delete Shared Block into its own component.
- Using `MenuGroup`s in the block settings menu introduces a bit of extra padding. It could be removed using a modifier. Let me know.

Here's a couple of screenshots showing how it looks now:
![screen shot 2018-06-27 at 5 26 49 pm](https://user-images.githubusercontent.com/677833/41966389-93b4cbd8-7a31-11e8-9c13-55afc4e2fe97.png)
![screen shot 2018-06-27 at 5 27 13 pm](https://user-images.githubusercontent.com/677833/41966407-9a95ab7a-7a31-11e8-82c0-5c69b1f05a06.png)

talldan commented:
Another very little thing I thought. It feels like when the menu is open, the ellipsis icon button should really not become invisible. In my screenshots above, the popover menu feels like its calling out from nowhere.

karmatosed commented:
This looks great and thanks for the hardwork wrangling this @talldan. I agree let's have the ellipsis show.

noisysocks commented:
> I've changed it now so that it applied to all MenuGroups. I had a search in the codebase, and the only thing that it might affect that I'm unsure how to view is something called PluginSidebarMoreMenuItem - I guess I'll need some plugins to view that?

You can test the plugins menu by pasting [this snippet](https://gist.github.com/gziolo/e2954aa83aa1f823b2b05ca1660c2223#file-my-plugin-es5-js) (the ES5 version) into your DevTools console. This will add a button into the More menu. See https://github.com/WordPress/gutenberg/pull/6031 for some context on that feature.

noisysocks commented on editor/components/block-settings-menu/shared-block-delete-button.js:
```diff
@@ -0,0 +1,62 @@
+import { isSharedBlock } from '@wordpress/blocks';
+import { withSelect, withDispatch } from '@wordpress/data';
+
+export function SharedBlockDeleteButton( { sharedBlock, onDelete, itemsRole } ) {
```
I like that these components only have one responsibility now 🙂🌈

noisysocks reviewed (changes requested):
Great work on this! It's looking really really close.

I'm not sure about re-using `MenuGroup` in this way. I support the general idea but there's a few implications:

1. We now have two `<div role="menu">`s nested within a `<div role="menu">` which feels a little weird to me.

<img width="663" alt="screen shot 2018-06-28 at 10 54 08" src="https://user-images.githubusercontent.com/612155/42007262-5b95e386-7ac2-11e8-9806-3d9b99dfc1c4.png">

2. The child `<div role="menu">`s have an `aria-labelledby` attribute which refers to an identifier that doesn't exist. Admittedly, this is probably a bug in `MenuGroup`.

3. Moving the `border-bottom` rule to `MenuGroup` has caused some regressions with `PluginSidebarMoreMenuItem`, like you suspected it would:

<img width="276" alt="screen shot 2018-06-28 at 10 56 16" src="https://user-images.githubusercontent.com/612155/42007323-b3567428-7ac2-11e8-8a71-f201e3762ed0.png">

4. The extra padding—like you said 🙂

---

I notice that we have a `.editor-block-settings-menu__separator` class defined already, so an alternative to `MenuGroup` might be to simply use a div with that class to add the separator line. You can see an example of us doing exactly that in `editor/components/block-settings-menu/block-transformations.js`.

talldan committed:
Revert "Apply border between MenuGroups"

This reverts commit c20aadacdfefc0b316352bfb61b35961fad2316d.

talldan committed:
Replace menu groups with a div that has a separator class applied

Using menu groups introduces a few issues in this menu
- They have extra padding by default
- They introduce another role=menu attribute
- the aria-labelledby attribute refers to a non-existant id

The separator class is already used for block transformations in this
menu so is a good fit.

talldan committed:
Rename SharedBlockSettings component to SharedBlockConvertButton

Since a recent refactor, this component now only renders buttons for
converting to and from shared blocks. This commit renames it to that
effect.

talldan committed:
Remove assertion already covered in snapshot test

hedgefield commented:
LOVE THIS.
